### ISSUE TITLE
Batch request creates a new http connection which invalidates the csrf-token

### ISSUE THREAD

**Issue by monczszilard** [object/Simple.OData.Client#263]
We are using a csrf-token concept. The token is valid only within the same http session. I get the token value via non-batch GET request and want to use it for a batch. But it seems that the batch creates a new http session where the token is not valid anymore.
Is there any way to reuse/share the existing ODataClient http session (connection) in the ODataBatch class (or vice versa)?

**Comment by object** [comment]
Hi, currently it's not possible: batch always creates a new instance of ODataClient object with a new session. But can you obtain the token from within the batch session and then reuse?

**Comment by monczszilard** [comment]
No, I cannot. To obtain a token i need a GET request as POST requests requires a valid token to be processed.
But anyway, could you provide an example how to make more batch requests using the same ODataBatch client instance? Can I simply reuse it - i.e. after calling the Exec... all actions are cleared and ready to define new ones?

**Comment by object** [comment]
Sure. Have a look at the following examples:

https://github.com/object/Simple.OData.Client/blob/master/Simple.OData.Client.Tests.Net40/BatchTypedTests.cs
https://github.com/object/Simple.OData.Client/blob/master/Simple.OData.Client.Tests.Net40/BatchTests.cs
https://github.com/object/Simple.OData.Client/blob/master/Simple.OData.Client.Tests.Net40/BatchDynamicTests.cs

and a few batch examples here:
https://github.com/object/Simple.OData.Client/blob/master/Simple.OData.Client.IntegrationTests/TripPinTests.cs

**Comment by akousmata** [comment]
@monczszilard did you end up coming up with a solution for this?  I'm in the exact same situation and am uncertain how to proceed.  I can get the x-csrf-token from a request to the metadata endpoint, but then I need to new up an ODataBatch object and no matter how I approach this problem, the token is always invalid.

**Comment by monczszilard** [comment]
@akousmata: Yes, I simply changed the source to share the existing connection from client to batch. I think these changes were needed to get that working:

```
       public ODataBatch(IODataClient client)
        {
            _client = new ODataClient(client as ODataClient, _entryMap);
        }

    internal ODataClient(ODataClient client, SimpleDictionary<object, IDictionary<string, object>> batchEntries)
        {
            _settings = client._settings;
            _session = client.Session;
            _requestRunner = client._requestRunner;

            if (batchEntries != null)
            {
                _batchEntries = batchEntries;
                _lazyBatchWriter = new Lazy<IBatchWriter>(() => _session.Adapter.GetBatchWriter(_batchEntries));
            }
        }
```

**Comment by akousmata** [comment]
Hmm, well, I was hoping for something that didn't involve changing the source.  @object any chance you would look into accepting something like as a pull request?  It must not be that common of a thing to do in the OData world, but x-csrf tokens quite common for security purposes in a lot of other libraries and online services.  In my they are a strict requirement as I am connecting to an SAP web service and the service providers are unwilling to accept batched requests without one.  If I don't make the suggested changes to the source code, I can't can't really utilize the ODataClient as is.

**Comment by akousmata** [comment]
@monczszilard don't you also need to clear out the request headers on subsequent requests?  I'm assuming you had to modify the request to include the `x-csrf-token: fetch` header in the initial request.  But then what did you do to update the header on the POST? I added a new function on the `ODataClient` to accomplish this, but wanted to see if you had a similar solution:

```
public void ChangeSettings(Action<ODataClientSettings> change)
{
    change(_settings);
}
```
Then called like this:

```
client.ChangeSettings(s =>
{
    // Reset the request headers
    s.BeforeRequest += (request) =>
    {
        if (request.Headers.Contains("x-csrf-token"))
            request.Headers.Remove("x-csrf-token");

        request.Headers.Add("x-csrf-token", token);
    };
    s.AfterResponse = null;
});
```

**Comment by object** [comment]
Let me investigate this further. Of course we should try to integrate this into the main source. If you have an opportunity to create a pull request, it will be highly appreciated.

**Comment by akousmata** [comment]
#396 this PR is different from my example above which gave a lot of control to the caller and instead scoped it to just updating the request headers.

**Comment by object** [comment]
Released in 4.26.

**Comment by akousmata** [comment]
Thanks for the quick turn around on this.
